src/courses/mappers/04.md

```diff
@@ -0,0 +1,342 @@
+---
+order: 4
+next: 05.md
+title: What Is an OHDF Mapper?
+author: Charles Hu
+---
+
+## What Is a Mapper?
+
+A mapper is a design pattern used to correlate (or map) items in two different objects with one another. Mappers are useful in that they allow us to connect items in objects that are nominally different but semantically the same. A result of this is that we can easily transform one object type into another through the use of mappings which define the direct relationship of items in each object type.
+
+Here are some scenarios which demonstrate some key aspects of mappers:
+
+::: details Transferring IDs
+Say we have to transfer the credentials of an employee who is moving from Business A to Business B. The data formats the businesses use for IDs are as follows:
+```json
+// Business A
+{Name, DoB, Age, Title}
+
+// Business B
+{employee, employeeBirthday, employeeAge, jobTitle}
+```
+
+How do we transfer John's credentials from Business A to Business B?
+```json
+{Name: 'John Doe', DoB: 10-6-1992, Age: 32, Title: 'Security Technician'}
+```
+
+What we can do is create a mapping which correlates the items from Business A's ID scheme to Business B's:
+```json
+{employee: Name, employeeBirthday: DoB, employeeAge: Age, jobTitle: Title}
+```
+
+With this, we can then develop a mapper which takes John's credentials from Business A and transforms it to Business B's format as so:
+```json
+{Name: 'John Doe', DoB: 10-6-1992, Age: 32, Title: 'Security Technician'}
+
+ ||
+\  /
+ \/
+
+MAPPER
+
+ ||
+\  /
+ \/
+
+{employee: 'John Doe', employeeBirthday: 10-6-1992, employeeAge: 32, jobTitle: 'Security Technician'}
+``` 
+
+The important thing to note here is that mappers rely on underlying mappings which match semantically similar fields between two objects. These matches allow us to correctly convert each item in one object to the other.
+:::
+
+::: details Translating a foreign language
+Say we are given the job to translate the traditional Chinese sentence '我愛貓' into English.
```

Comment:
I think we need a lot more explanation here. In your example, without know the language grammar being translated, it is impossible to end with the final result, there is, go from `cats I love` to `I love cats`  

I would state that a mapper like in linguistics is the grammar that is used to bridge one objects content to another. It is used to analyze the syntax and apply the necessary morphology.

Somewhere, perhaps in the beginning of defining a mapper we need to make this obvious, there is, the more definitions (grammar) that is added to a mapper the better the translation/transformation between the objects and a higher the fidelity can be obtained.

Reply:
I've removed that example specifically since it has become clear that it'll bring up questions of linguistics and grammar more than mappers, which is a mess that I don't really want to delve into. I've noted some of your points in the rewrite of this section though.